### PR TITLE
Add support for DateTime fields and optimize search queries performance

### PR DESCRIPTION
**Pull request by lebaz20**
### Part I Add support for field type DateTime to be used in POST, PUT, ..etc requests
- Current state is that the field value is turned into array with key 'id' holding the value, which is not accepted according to field type ,Which is DateTime ,Throwing corresponding exception

### Part II Optimize search and retrieving queries performance when not relying on criteria, Just limit, orderBy, ...etc
- Include limit and offset into the main query, To get the needed portion only
- Count objects through count query ,Not regular query
- Very obvious for extremely large tables
- Test results : 
 - Table size: 136740 rows
 - Query: _limit = 1
 - Time before optimization : 1h 9m ,and query never finished 'Instead memory was exhausted, even after allowing 256MB for Apache'
 - Time after optimization : 9s

### CODE REVIEW

**Comment by stanlemon** [comment]
@lebaz20 It looks like TravisCI was having a bad day when the build ran, I'll rerun and try to follow up with you this weekend.  My apologies for the delay, I've been traveling as of late.

File: Object/Manager.php [hunk 1]
```diff
@@ -75,20 +75,30 @@ public function search(Criteria $criteria)
...
+            $whereConditions = $expressionBuilder->andX($whereConditions, $expressionBuilder->eq("e.{$key}", $value));
         }
 
-        $allObjects = $this->getRepository()->findBy(
-            $criteria->toArray(),
-            $criteria->getOrderBy()
+        $filteringCriteria = $criteria->toArray();
+        $objects = $this->getRepository()->findBy(
+            $filteringCriteria,
+            /*$orderBy =*/$criteria->getOrderBy(),
```
Comment (stanlemon) [inline Object/Manager.php]: Don't need to keep these queries.

**Comment by stanlemon** [comment]
@lebaz20 I was able to verify existing tests do not break, I'm not sure why travisci isn't playing nice.  Can you please add additional test coverage for the functionality you are adding? Once there is added test coverage I will merge these changes in. Thanks for your contribution!

**Comment by stanlemon** [comment]
@lebaz20 In addition to the new test coverage, some of the mongodb tests are now failing and will need to be fixed:

```
There was 1 error:

1) Lemon\RestBundle\Tests\Controller\MongoResourceControllerTest::testListAction
PHPUnit_Framework_Exception: Argument #2 (No Value) of PHPUnit_Framework_Assert::assertCount() must be a countable or traversable

/Users/stan/Sites/lebaz20-rest-bundle/Tests/Controller/ResourceControllerTest.php:80

--

There was 1 failure:

1) Lemon\RestBundle\Tests\Controller\MongoResourceControllerTest::testInvalidMethod
Failed asserting that null matches expected 2.

/Users/stan/Sites/lebaz20-rest-bundle/Tests/Controller/ResourceControllerTest.php:1321
```

**Comment by lebaz20** [comment]
I am already working on a fix, i will make sure code coverage is maintained,

File: Object/Manager.php (hunk 1 above)
Comment (lebaz20) [inline Object/Manager.php]: This is the main query, There is another query for the count, But this is the main one with the criteria, order, limit and offset

**Comment by lebaz20** [comment]
@stanlemon Sorry for being that late, I fixed all unit test failures, revised unit test coverage, and added support for MongoDB

**Comment by stanlemon** [comment]
@lebaz20 Thank you for submitting this! Work has me super swamped, I am hoping to take a closer look over the Holiday next week and get this merged in. Sorry for the delay, but thanks so much for the contribution!

**Pull request merged**